pkg/loadbalancer/nfqlb/nfqlb.go

```diff
@@ -17,152 +17,305 @@ limitations under the License.
+	return err
+}
+
+// Delete -
+// Delete decreases reference count of the file backing the shared mem lb
+//
+// Notes:
+// - The file is not removed by the OS unless no other references are held
+// - Flow rule also store references towards the shared mem lb it's associated with.
+//   So they must be removed as well to get rid of a shared mem lb. (But that is
+//   the responsiblility of the "user" i.e. the Stream construct.)
+// - Previously activated Targets are not deactivated, as that information shall
+//  disappear once the shared mem is destroyed.
+func (n *nfqlb) Delete() error {
+	ctx := context.TODO()
+	// unlink the shared mem file
+	cmd := exec.CommandContext(
+		ctx,
+		nfqlbCmd,
+		"delete",
+		fmt.Sprintf("--shm=%s", n.name),
+	)
+
+	logrus.Infof("Delete nfqlb shared mem lb: %v", cmd.String())
+	stdoutStderr, err := cmd.CombinedOutput()
 	if err != nil {
-		return nil, err
+		err = fmt.Errorf("%v; %s", err, stdoutStderr)
 	}
-	return lb, nil
+	return err
 }
 
+// Activate -
+// Activate adds a Target with the very identifier to the associated shared mem lb
 func (n *nfqlb) Activate(identifier int) error {
-	var stderr bytes.Buffer
-	cmd := exec.Command(
+	ctx := context.TODO()
+	stdoutStderr, err := exec.CommandContext(
+		ctx,
 		nfqlbCmd,
 		"activate",
 		fmt.Sprintf("--index=%d", identifier-1),
-		fmt.Sprintf("--shm=%s", n.getTargetSHM()),
+		fmt.Sprintf("--shm=%s", n.name),
 		strconv.Itoa(identifier),
-	)
-	cmd.Stderr = &stderr
-	_, err := cmd.Output()
+	).CombinedOutput()
 	if err != nil {
-		return fmt.Errorf("%w; %v", err, stderr.String())
+		err = fmt.Errorf("%v; %s", err, stdoutStderr)
 	}
-	return nil
+	return err
 }
 
+// Deactivate -
+// Deactivate removes a Target with the very identifier from associated shared mem lb
 func (n *nfqlb) Deactivate(identifier int) error {
-	var stderr bytes.Buffer
-	cmd := exec.Command(
+	ctx := context.TODO()
+	stdoutStderr, err := exec.CommandContext(
+		ctx,
 		nfqlbCmd,
 		"deactivate",
 		fmt.Sprintf("--index=%d", identifier-1),
-		fmt.Sprintf("--shm=%s", n.getTargetSHM()),
-	)
-	cmd.Stderr = &stderr
-	_, err := cmd.Output()
+		fmt.Sprintf("--shm=%s", n.name),
+	).CombinedOutput()
 	if err != nil {
-		return fmt.Errorf("%w; %v", err, stderr.String())
+		err = fmt.Errorf("%v; %s", err, stdoutStderr)
 	}
-	return nil
+	return err
 }
 
-// https://github.com/Nordix/nfqueue-loadbalancer/blob/master/src/nfqlb/cmdLb.c#L162
-func (n *nfqlb) Start() error {
-	n.cmd = exec.Command(
+// SetFlow -
+// SetFlow adds a Flow selecting the associated shared mem lb representing a Stream
+//
+// Note:
+// It also weeds out /0 IP ranges and 'any' port filters if possible to simplify the
+// config and improve match performance.
+func (n *nfqlb) SetFlow(flow *nspAPI.Flow) error {
+	ctx := context.TODO()
+	args := []string{
+		"flow-set",
+		fmt.Sprintf("--name=%v-%v", n.name, flow.GetName()),
+		fmt.Sprintf("--target=%v", n.name),
+		fmt.Sprintf("--prio=%v", flow.GetPriority()),
```

Comment:
You _never_ have to delete a flow to re-define is with any other values. You should just do a `flow-set` with the updated values. `nfqlb` will grab the write-lock and sort the flows. The operation is atomic, but it still takes time and since the write-lock is held packet handling is suspended. In very heavy traffic cases that may cause nfqueue losses, but I don't think that will happen.

Reply:
Okay, thank you, just asking because with nftables priority update was not possible (had to delete and re-create the rule). And this is blocked in the Operator, but now we could change it.